src/Algebra/Properties/Magma.agda

```diff
@@ -0,0 +1,75 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Some items and theory for Magma
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+open import Algebra using (Magma)
+open import Algebra.Definitions using (Commutative; LeftQuotient)
+open import Data.Product using (_×_; _,_)
+open import Function using (_∘_)
+open import Level using (_⊔_)
+open import Relation.Binary using (Rel; _Respects_; _Respects₂_)
+open import Relation.Nullary using (¬_)
+
+module Algebra.Properties.Magma {a ℓ} (M : Magma a ℓ) where
+
+open Magma M
+
+------------------------------------------------------------------------------
+-- The divisibility relation
+------------------------------------------------------------------------------
+
+infix 5 _∣_ _∤_ _∣∣_ _¬∣∣_
+
+private C = Carrier
+
+_∣_ : Rel C (a ⊔ ℓ)                   -- x ∣ y  denotes  x divides y
+x ∣ y =  LeftQuotient _≈_ _∙_ y x     -- as  y ≈ q ∙ x  for some q.
```

Comment:
These definitions don't depend on the magma axioms so I think it would make more sense for the contents of this file to live in it's own file (`Algebra.Divisibility`?) parameterised by the equality relation and operator. The proofs will then be parameterised by the exact properties they need (e.g. transitivity, congruency). This will make the whole development much more flexible to use.

Reply:
Please, consider this once more. If you insist, then I would change this.
Meanwhile consider the following.
``_∣_`` needs all the relations and operations of Magma ``(_≈_, _∙_), and the proofs for  ``_∣_`` in this module also use
`` ∙-cong`` and the properties of ``≈``, both  coming from Magma.
So that ``_∣_`` needs exactly the whole environment of Magma.
And if in such an evident situation we do not use a parameter of Magma, then why Magma is at all introduced?
Ii is introduced exactly for the aim of replacing the environment of several relations and operation with a single word of Magma.  For the same reason other classical categories are introduced: Semigroup, Group, Ring etc.

Reply:
Divisibility has sense only in Magma. So, Magma needs to be the parameter type of the module.
In textbooks, divisibility is most often defined for Ring (for its multiplication), which is rather restrictive.  In our design, divisibility in a  (Semi)ring R  will be imported from Algebra.Properties.Magma (Semiring.*-magma R).
Please, agree with this.

Reply:
```
>>  x ∣ y = ∃ (λ q → y ≈ q ∙ x)       -- Def
``` 
> These definitions don't depend on the magma axioms so I think it would make more sense for the contents > of this file to live in it's own file (Algebra.Divisibility?) parameterised by the equality relation and operator. 

Now, I doubt.

Do you expect the definition ``Def`` to be used in the environment that does not include the Magma laws? 
I do not know, may be you think of RawMagma and "raw-divisibility" _|_ that can be used in provers?
Can they?
(I do not recall, why provers may tend to use Raw<structure> instead  of <structure>).

1) ``Def`` depends on all the Magma _signature_ (``Carrier, _≈_, _∙_``). 
2) I do not expect this _∣_ to be used for `≈` being not an equivalence, or for _∙_ being not congruent.
    If it is used so, then it would be called differently, say, pseudo-divisibility, and would have another identifier.
3) Mathematicians always suppose that a relation  ``_∣_``  called _divisibility_ needs to satisfy reflexivity and
    transitivity. And their proof already requires the laws of Monoid. So that the parameter Magma is already    
    a very generic environment. 
4) The "Raw" definitions complicate and increase the code.

?

Reply:
> Do you expect the definition Def to be used in the environment that does not include the Magma laws? I do not know, may be you think of RawMagma and "raw-divisibility" | that can be used in provers?
 
There's certainly solver theories that involve divisibility out there. That's exactly the point: who knows how this code will be used in the future? Given that, it makes sense to ensure the code is as modular and reusable as possible. This approach is followed across the library.

> The "Raw" definitions complicate and increase the code.

The proofs would then be re-exported and correctly parameterised in `Algebra.Properties.(Magma/Semigroup/Monoid)`. This means that the public facing API for the users of the library remains unchanged.

Reply:
Please, consider my commit of October 26, 2020.
All your notes are implemented - except few ones to which I provide questions.
```
>>  x ∣ y = ∃ (λ q → y ≈ q ∙ x)       -- Def
```
> These definitions don't depend on the magma axioms so I think it would make more sense for the contents 
> of this file to live in it's own file (Algebra.Divisibility?) parameterised by the equality relation and operator.

So far, I put it to  ``Algebra.Definitions`` as
```
module Divisibility (_∙_ : Op₂ A)                                                 
  where                                                                           
  infix 5 _∣_ _∤_ _∣∣_ _¬∣∣_                                                      
                                                                                  
  _∣_ : Rel A (a ⊔ ℓ)           -- x ∣ y  denotes  x divides y                    
  x ∣ y = ∃ (λ q → y ≈ (q ∙ x))                                                   
  ...
```
Respectively its import to other modules in Algebra/Properties has changed. For example,
 ```
...
open import Algebra.Definitions using (Commutative; module Divisibility)          
...                                                                                   
module Algebra.Properties.Magma {a ℓ} (M : Magma a ℓ) where                       
                                                                                  
open Magma M                                                                      
...                                                                                
open Divisibility {A = Carrier} _≈_ _∙_ public using (_∣_; _∤_; _∣∣_; _¬∣∣_)      
                                                                                  
∣-respˡ : ∀ {y} → (_∣ y) Respects _≈_                                             
∣-respˡ x≈x' (q , y≈qx) = q , trans y≈qx (∙-congˡ x≈x')                           
...
```          
I am going to convince you that this is appropriate.
Divisibility is not such an outstanding definition as to take a separate file. It is not any more remarkable than, say,
`LeftInverse`. It is similar to several definitions in ``Algebra.Definitions``. For example,
```
RightIdentity : A → Op₂ A → Set _                                                 
RightIdentity e _∙_ = ∀ x → (x ∙ e) ≈ x
```
defines what is a right identity with respect to  `≈` and  ``_∙_``.
Similarly
```
Divides : Op₂ A → Rel A _
Divides _∙_ x y = ∃ (λ q → y ≈ (q ∙ x)) 
```
could define what does it mean "x divides y relatively to `≈` and  ``_∙_``".
So, is this an appropriate place to set `Divides` ?
Further, we need to use an infix ``_∣_`` instead of divides. But how to agree the infix syntax for  ``_∣_`` with the third argument ``_∙_`` which is not in the parameters of ``Algebra.Definitions`` ?  
I do not find any other way as to introduce the above ``module Divisibility``, with declaring  ``_∣_`` inside it.
Will this do?